📖  Add MHCClass spec and details to proposal (#5511)
kubernetes-sigs/cluster-api: docs/proposals/202105256-cluster-class-and-managed-topologies.md

**Notes**: Same notes as in Story 4 apply.

#### Story 8 - Easy UX for MachineHealthChecks
As a cluster operator I want a simple way to define checks to manage the health of the machines in my cluster. 

Instead of defining MachineHealthChecks each time a Cluster is created, there should be a mechanism for creating the same type of health check for each Cluster stamped by a ClusterClass.

**Examples**: 

> - A cluster operator would like to create a single MachineHealthCheck that applies across each machineDeployment and controlPlane created for Clusters stamped by a ClusterClass.

Reviewer: ```suggestion
- A cluster operator would like to create a single MachineHealthCheck that applies across each MachineDeployment and ControlPlane created for Clusters stamped by a ClusterClass.
```

> - A cluster operator would like to have individual MachineHealthChecks for each machineDeployment in Clusters stamped from a ClusterClass.

Reviewer: ```suggestion
- A cluster operator would like to have individual MachineHealthChecks for each MachineDeployment in Clusters stamped from a ClusterClass.
```

**Notes**: Same notes as in Story 4 apply.

#### Story 8 - Easy UX for MachineHealthChecks
As a cluster operator I want a simple way to define checks to manage the health of the machines in my cluster. 

Instead of defining MachineHealthChecks each time a Cluster is created, there should be a mechanism for creating the same type of health check for each Cluster stamped by a ClusterClass.

**Examples**: 

> - A cluster operator would like to create a single MachineHealthCheck that applies across each MachineDeployment and ControlPlane created for Clusters stamped by a ClusterClass.

Reviewer 2: In most cases, we should try to separate the control-plane MHC from Machine Deployment ones, they might also have different defaults. For example, each control plane node might take longer to come up than worker nodes.

Author: To start with I think it's simpler to reconcile all defaults to the same - with the current webhook defaults providing most of the values. Later if we have a strong idea of how the defaults should differ between control planes and MHCs we can have multiple defaults.

Reviewer 2: The control plane nodes are usually always different than the others, for example a control plane might take a lot longer to join that a worker node. We should probably avoid an overall set of parameters that's shared for all of the nodes, including control plane ones; from a user perspective, I'd expect that a healthCheck section under Control Plane to only apply to control plane nodes rather, and one under workers to apply to all workers instead.

Thoughts?

cc @CecileRobertMichon @fabriziopandini

Reviewer 3: +1 to rephrase this; if fact we are planning to create 1 MHC for the CP and 1 for each MachineDeployment (vs having a single MachineHealthCheck)

Author: Yeah this examples section is out of sync. I've removed it completely. It was added to show how that use case - global MHC definition - would work, but it's no longer in the proposal.

  // Note: Patches will be applied in the order of the array.
  // +optional
  Patches []ClusterClassPatch `json:"patches,omitempty"`
  
  // MachineHealthCheckClass defines a MachineHealthCheck for each 

>   // ControlPlaneClass and MachineDeploymentClass in this ClusterClass.

Reviewer 2: Which one we've decided to go for?

>   // ControlPlaneClass and MachineDeploymentClass in this ClusterClass.

Author: This is exclusive in this proposal - there's an issue #5770 to decide whether to add overriding down the line.

  // +optional

>   MachineHealthCheck  *MachineHealthCheckClass `json:"machineHealthCheck,omitempty"`

Reviewer: I think we should be specific here and keep it as is. 

I think we can always add other features with additional individual fields and if we want to group them at some point we can deprecate + merge.

For now it would be very speculative as we don't know how/if other features are added. 
* ClusterResourceSet seems to be out-of-scope of ClusterClass at the moment (as it's orthogonal to ClusterClass and applies across multiple clusters)
* Autoscaler: 
    * labels can be already added today with our metadata fields
    * deploying the autoscaler itself seems to rather fit into ClusterAddons bucket (https://github.com/kubernetes-sigs/cluster-api/issues/5491)

Overall ClusterClass itself is all about the cluster lifecycle, so I'm not sure an additional field/level like `integrations` or `lifecycleIntegrations` would add a lot of benefit and depending on which and how we add additional features the field name might not fit.

@elmiko WDYT?

>   MachineHealthCheck  *MachineHealthCheckClass `json:"machineHealthCheck,omitempty"`

Reviewer 4: @sbueringer i think your reasoning makes good sense. i would be happy to see the integration defined better in an enhancement that comes out of #5491.

>   MachineHealthCheck  *MachineHealthCheckClass `json:"machineHealthCheck,omitempty"`

Reviewer 2: Could we move this definition under `WorkersClass`? And specify that this one will be applied to each one of the MachineDeployment classes unless they supply their own configuration/overrides?

>   MachineHealthCheck  *MachineHealthCheckClass `json:"machineHealthCheck,omitempty"`

Author: I think the most straightforward implementation is to define this on the lowest levels. If we want overriding behaviour later on we can introduce the higher level workerClass MHC.

>   MachineHealthCheck  *MachineHealthCheckClass `json:"machineHealthCheck,omitempty"`

Reviewer 2: Having the MHC Class top level can be confusing to users, because it's one level above ControlPlane so folks might think that there is some layering in place.

>   MachineHealthCheck  *MachineHealthCheckClass `json:"machineHealthCheck,omitempty"`

Reviewer 3: +1 to have MHC only inside CP and MD for the first iteration (and thus to remove it from here)

}

// ControlPlaneClass defines the class for the control plane.
type ControlPlaneClass struct {
  Metadata ObjectMeta `json:"metadata,omitempty"`

  // LocalObjectTemplate contains the reference to the control plane provider.
  LocalObjectTemplate `json:",inline"`

  // MachineTemplate defines the metadata and infrastructure information
  // for control plane machines.
  //
  // This field is supported if and only if the control plane provider template
  // referenced above is Machine based and supports setting replicas.
  // +optional
  MachineInfrastructure *LocalObjectTemplate `json:"machineInfrastructure,omitempty"`
  
  // MachineHealthCheckClass defines a MachineHealthCheck for this ControlPlaneClass.
  // +optional

>   MachineHealthCheck *MachineHealthCheckClass `json:"machineHealthCheck,omitempty"`

Reviewer: Yup, let's discuss it only in: https://github.com/kubernetes-sigs/cluster-api/pull/5511/files#r738769812

  // Note: Patches will be applied in the order of the array.
  // +optional
  Patches []ClusterClassPatch `json:"patches,omitempty"`
  

>   // MachineHealthCheckClass defines a MachineHealthCheck for each 
>   // ControlPlaneClass and MachineDeploymentClass in this ClusterClass.
>   // +optional

Reviewer 2: ```suggestion
  // MachineHealthCheck defines the default MachineHealthCheck configuration 
  // used when creating a ControlPlane and for each MachineDeployment in this ClusterClass.
  // +optional
```

Author: This isn't quite right - there should only be a minimal difference between the defaulting in ClusterClass MHCs vs the defaulting in the webhook see #5871.

Reviewer 2: By defaulting I meant that these are the parameters used when creating the MHC object

Author: This structure is out of sync with the proposal. I've removed it - MHCs should only be defined under the ControlPlane or under the MachineDeployments.

  // Template is a local struct containing a collection of templates for creation of
  // MachineDeployment objects representing a set of worker nodes.
  Template MachineDeploymentClassTemplate `json:"template"`
  
  // MachineHealthCheckClass defines a MachineHealthCheck for this MachineDeploymentClass.
  // +optional
  MachineHealthCheck *MachineHealthCheckClass `json:"machineHealthCheck,omitempty"`
}

> [// MachineDeploymentClassTemplate defines how a MachineDeployment generated from a MachineDeploymentClass

Reviewer 3: ```suggestion
// MachineDeploymentClassTemplate defines how a MachineDeployment generated from a MachineDeploymentClass
```

// should look like.

> ]()type MachineDeploymentClassTemplate struct {

Reviewer 3: ```suggestion
type MachineDeploymentClassTemplate struct {
```

  // Template is a local struct containing a collection of templates for creation of
  // MachineDeployment objects representing a set of worker nodes.
  Template MachineDeploymentClassTemplate `json:"template"`
  

>   // MachineHealthCheckClass defines a MachineHealthCheck for this MachineDeploymentClass.

Reviewer: What was the resolution of the following?
> Could we move this definition under WorkersClass? And specify that this one will be applied to each one of the MachineDeployment classes unless they supply their own configuration/overrides?
> https://github.com/kubernetes-sigs/cluster-api/pull/5511#discussion_r766768674?

(Although I think @vincepri meant to copy it there so there is `machineHealthCheck` on workers and on the MD level)

>   // +optional

Reviewer: ```suggestion
  // This field is supported if and only if the control plane provider template
  // referenced above is Machine based and supports setting replicas.
  // +optional
```

Q: Does this restriction of the `machineInfrastructure` also apply to `machineHealthCheck`? 
Q2: If yes, what do we need Machine based **and** replicas, or only Machine based?

Reviewer: Based on the implementation PR I would assume we have the same requirements (Machine based and replicas)

Reviewer 2: Is the comment in the wrong spot? Isn't the field above the MachineDeployment one?

Reviewer: Absolutely, probably mis-mapped it when reviewing via Intellij

Author: Discussed this with @sbueringer and I've added the above comment into L258-L260 i.e. in the MachineHealthCheckClass for the ControlPlane

  // Template is a local struct containing a collection of templates for creation of
  // MachineDeployment objects representing a set of worker nodes.
  Template MachineDeploymentClassTemplate `json:"template"`
  
  // MachineHealthCheckClass defines a MachineHealthCheck for this MachineDeploymentClass.
  // +optional

>   MachineHealthCheck *MachineHealthCheckClass `json:"machineHealthCheck,omitempty"`

Reviewer: Yup, let's discuss it only in: https://github.com/kubernetes-sigs/cluster-api/pull/5511/files#r738769812

  // offered by a provider.
  Ref *corev1.ObjectReference `json:"ref"`
}

// MachineHealthCheckClass defines a MachineHealthCheck for a group of Machines.
type MachineHealthCheckClass struct {
  // UnhealthyConditions contains a list of the conditions that determine
  // whether a node is considered unhealthy. The conditions are combined in a

>   // logical OR, i.e. if any of the conditions is met, the node is unhealthy.

Reviewer 3: Should we add a note about this field being defaulted when the actual MHC will be created?

Author: Added a note on defaulting above the struct.

  // offered by a provider.
  Ref *corev1.ObjectReference `json:"ref"`
}

// MachineHealthCheckClass defines a MachineHealthCheck for a group of Machines. 
// Note: all fields are defaulted on object creation through the MachineHealthCheck defaulting webhook with the exception

> // of UnhealthConditions. UnhealthyConditions is defaulted during reconciliation if it has not been profided by the user.

Reviewer 5: ```suggestion
// of UnhealthConditions. UnhealthyConditions is defaulted during reconciliation if it has not been provided by the user.
```

Reviewer 5: could you expand on why `UnhealthyConditions` is special and is defaulted in the controller and not in a webhook?

Author: I've updated the comments on the struct. The difference is that the webhook defaults all fields other than UnhealthyConditions, and it's a required field. To avoid needing a breaking change like #5871 the idea here is to default in reconciliation.

Reviewer 3: UX is also another reason behind defaulting; this allows to get an opinionated MHC with sane defaults by specifying an empty object

Author: The proposal now has no special defaulting for MHC - users will be required to define UnhealthConditions in each place they define an MHC inside the ClusterClass.

  // whether a node is considered unhealthy. The conditions are combined in a
  // logical OR, i.e. if any of the conditions is met, the node is unhealthy.

>   // Note: in order to provide a simpler UX, in case the user does not provide a value for this field, The topology reconciler applies

Reviewer: Why do we "hide" the defaulting in the reconcile loop here instead of defaulting on the ClusterClass webhook? (like we afaik usually do) I think defaulting it in the ClusterClass webhook level would be more transparent for users (if we want to do it).

Also (applies to both variants) when we treat an empty `UnhealthyConditions` array like it has our default value, it seems to be impossible to configure a ClusterClass to not use `UnhealthyConditions`? Is that a problem?

Reviewer: Just learned from Killian that the condition array is not allowed to be empty. So as an empty array is an invalid configuration my second point isn't a problem.

Author: The latest version no longer performs any defaulting on the MHC in the reconciler and leaves all defaulting the same as for vanilla MHC

  // offered by a provider.
  Ref *corev1.ObjectReference `json:"ref"`
}

// MachineHealthCheckClass defines a MachineHealthCheck for a group of Machines. 

> 

Reviewer 3: ```suggestion
```

  // offered by a provider.
  Ref *corev1.ObjectReference `json:"ref"`
}

// MachineHealthCheckClass defines a MachineHealthCheck for a group of Machines.

> type MachineHealthCheckClass struct {

Reviewer 5: The fields below seem to be mostly duplicates of the fields in `MachineHealthCheckSpec`. Are we concerned this might cause some trouble long-term for maintenance? How do we ensure the two don't diverge? What if a new feature is added to one place but not the other? Have we considered sharing a common struct so they can share the fields that are in both places or making one a subset of the other?

Please let me know if this comment feels out of scope. I realize this is a pattern that exists in other places in ClusterClasses, which is why this may be a bigger discussion.

Reviewer 3: +1 to start working on some guidelines for common structs, given that this was discussed also in last week office hours

  // offered by a provider.
  Ref *corev1.ObjectReference `json:"ref"`
}

// MachineHealthCheckClass defines a MachineHealthCheck for a group of Machines.
type MachineHealthCheckClass struct {
  // UnhealthyConditions contains a list of the conditions that determine

>   // whether a node is considered unhealthy.  The conditions are combined in a

Reviewer: ```suggestion
  // whether a node is considered unhealthy. The conditions are combined in a
```
nit

  // logical OR, i.e. if any of the conditions is met, the node is unhealthy.

>   // If not set this field will default to an array of two values:
>   //     - type: Ready
>   //       status: Unknown
>   //       timeout: 300s
>   //     - type: Ready
>   //       status: "False"
>   //       timeout: 300s

Reviewer: ```suggestion
```
Based on that we're deferring the defaulting discussion to a separate issue, I think we should remove this comment for now.

Reviewer: I would also remove l.332 and l.347 for now. As long as those fields are only defaulted on the MHC itself I think we shouldn't mention it here. Having the godoc here very strongly applies that this exact fields are defaulted on ClusterClass creation/update.

  // +optional

>   UnhealthyConditions []UnhealthyCondition `json:"unhealthyConditions,omitempty"`

Reviewer 3: ok, makes sense

>   UnhealthyConditions []UnhealthyCondition `json:"unhealthyConditions,omitempty"`

Reviewer: Sounds good to me two. We can eventually deprecate both fields at the same time, which seems okay.

  // ValueFrom defines the value of the patch.
  // Note: Either Value or ValueFrom is required for add and replace
  // operations. Only one of them is allowed to be set at the same time.

>   // +optionalhttps://github.com/kubernetes-sigs/cluster-api/pull/5511

Reviewer 3: ```suggestion
  // +optional
```

              the template and validating if the result is a valid YAML or JSON value.
    - We will only implement syntax validation. A semantic validation is not possible as the patches will be applied 
      to provider templates. **It’s the responsibility of the ClusterClass author to ensure the patches are semantically valid**.
  - `MachineHealthCheck`

>     - If defined at `ClusterClass.spec.machineHealthCheck` must not be defined in `ClusterClass.spec.workers.machineDeployments[i].machineHealthCheck` or `ClusterClass.spec.controlPlane.machineHealthCheck` 

Reviewer 3: ```suggestion
```
Given that we have MHC only at the leaf level

Author: Should be updated now.

              the template and validating if the result is a valid YAML or JSON value.
    - We will only implement syntax validation. A semantic validation is not possible as the patches will be applied 
      to provider templates. **It’s the responsibility of the ClusterClass author to ensure the patches are semantically valid**.
  - `MachineHealthCheck`

>     - (defaulting) All fields required for MachineHealthChecks, including UnhealthyConditions, are defaulted to known good values on creation.

Reviewer 3: Let's drop this given that we are not defaulting on ClusterClalss creation, but we are defaulting on object creation (might be worth adding a note down below)

              the template and validating if the result is a valid YAML or JSON value.
    - We will only implement syntax validation. A semantic validation is not possible as the patches will be applied 
      to provider templates. **It’s the responsibility of the ClusterClass author to ensure the patches are semantically valid**.

>   - `MachineHealthCheck`

Reviewer: I think it's intuitive when the "local" ClusterClass.spec.controlPlane.machineHealthCheck overrides the global ClusterClass.spec.machineHealthCheck (similar for MD).

We have kind of a similar pattern where we prioritize template-specific variables over global variables. (it is implemented in that way but it's current not used as the only template-specific variables are builtin variables which are disjunct from the global builtin variables)

P.S. I would expect a local machineHealthCheck to entirely overwrite a global one. I.e. not on a per-field basis, because then you would have to figure out how to unset global MHC fields and this would lead to a weird UX where you might be forced to unset global MHC fields per MD/ControlPlane.

@enxebre @killianmuldoon @ykakarap WDYT?

>   - `MachineHealthCheck`

Reviewer 3: I think we should keep things simple in the first iteration and keep global health check mutually exclusive to health checks at CP/MD level.

We can then relax this constraint at a later stage and introduce merge/override rules in a compatible way (while instead it will be a breaking change to remove it).

>   - `MachineHealthCheck`

Author: I've added an issue to cover the idea of an overriding MHC for ClusterClass at https://github.com/kubernetes-sigs/cluster-api/issues/5770

              the template and validating if the result is a valid YAML or JSON value.
    - We will only implement syntax validation. A semantic validation is not possible as the patches will be applied 
      to provider templates. **It’s the responsibility of the ClusterClass author to ensure the patches are semantically valid**.
  - `MachineHealthCheck`
    - If defined at `ClusterClass.spec.machineHealthCheck` must not be defined in `ClusterClass.spec.workers.machineDeployments[i].machineHealthCheck` or `ClusterClass.spec.controlPlane.machineHealthCheck` 

>     - (defaulting) All fields required for MachineHealthChecks, including UnhealthyConditions, are defaulted as described in the API definition.

Reviewer: I think I would prefer having the same defaulting on the ClusterClass webhook as on the MHC webhook to have the MHC config in the ClusterClass behave the same way as the MHC itself.

Some context:
* We currently default:
  * spec.maxUnhealthy not set => 100%
  * spec.nodeStartupTimeout not set => 10m
  * spec.remediationTemplate != nil && spec.remediationTemplate.namespace empty => spec.remedationTemplate.namespace => MHC.namespace

**If the defaulting behavior is the same**

* nodeStartupTimeout not set => defaulted in the ClusterClass => explicitly set in the MHC
    * => ClusterClass.spec.machineHealthCheck and MHC will be the same

**if we don't have defaulting in the ClusterClass**

* nodeStartupTimeout not set => not defaulted in the ClusterClass, but defaulted in MHC
    *  => MHC will have the timeout, ClusterClass.spec.machineHealthCheck won't
        * we have to implement a logic where we only keep ClusterClass.spec.machineHealthCheck sync with MHC if the fields in ClusterClass.spec.machineHealthCheck is not nil.

**What happens if the MHC defaulting changes**

We have the weird behavior that the MHC spec depends on when it has been created by the cluster topology controller, because the fields are "unmanaged" from a topology controller perspective. I.e. old MHC's will have the fields filled by the old default logic, new MHCs by the new. Depending on how the defaulting changes we might end up with a mixture.

When the ClusterClass itself defaults the defaults are persisted at the time of ClusterClass creation in etcd and are then consistently used.

tl;dr I lean towards keeping the defaulting logic sync (like we did in cases like KCP vs KCPTemplate) to avoid weird effects.

>     - (defaulting) All fields required for MachineHealthChecks, including UnhealthyConditions, are defaulted as described in the API definition.

Reviewer 3: > If the MHC controller takes care of handling the default behavior ... become a hidden behavior for the user.

I don't think this is a hidden behavior, but a feature of ClusterClass explicitly documented in [template<->object-reconciliation](https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/202105256-cluster-class-and-managed-topologies.md#notes-on-template---object-reconciliation).

ClusterClass as of today takes care of enforcing a _subset of fields_ in the generated objects, but the controller/the users are responsible for the missing ones, thus opening up to things like user customization or the controllers to store some information in the spec (e.g. for making them survive move operations).

And if you think about it is the same behaviour users get when they use kubectl apply (all the things the were explicit about, are preserved, everything else is delegated to the system that eventually apply defaults when things are created)

Introducing explicit defaulting for MHC will mean that ClusterClass, kind of goes in the opposite direction, so I'm -1 for this

>     - (defaulting) All fields required for MachineHealthChecks, including UnhealthyConditions, are defaulted as described in the API definition.

Reviewer: > Introducing explicit defaulting for MHC will mean that ClusterClass, kind of goes in the opposite direction

From that point of view it makes sense. So the main idea of ClusterClass in that respect is to preserve user intent (and don't overwrite MHC webhook intent). 

That means we should be very very conservative with any defaulting we do on ClusterClass. But this fits the current behavior as we're only defaulting namespaces right now.

>     - (defaulting) All fields required for MachineHealthChecks, including UnhealthyConditions, are defaulted as described in the API definition.

Author: I've created a parallel issue for defaulting at https://github.com/kubernetes-sigs/cluster-api/issues/5769 and removed the specifics of where defaulting will happen from the proposal to make it an implementation detail.

>     - (defaulting) All fields required for MachineHealthChecks, including UnhealthyConditions, are defaulted as described in the API definition.

Reviewer: Sounds good, can be resolved for now from my side.

          apiVersion: controlplane.cluster.x-k8s.io/v1alpha4
          kind: KubeadmControlPlaneTemplate
          name: vsphere-prod-cluster-template-kcp

>         # This will create a health check for each deployment created with the "linux-worker" MachineDeploymentClass

Reviewer: Let's be explicit with the resource name:
```yaml
# This will create a MachineHealthCheck for "vsphere-prod-cluster-template-kcp" KubeadmControlPlane machines.
```

                apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
                kind: VSphereMachineTemplate
                name: linux-vsphere-template

>           # This will create a health check for each deployment created with the "linux-worker" MachineDeploymentClass

Reviewer: ```suggestion
          # This will create a MachineHealthCheck for each deployment created with the "linux-worker" MachineDeploymentClass
```
nit, just to keep it consistent with the comment above

>           # The machineHealthCheck values will all be defaulted as no values are specified in the empty struct below.

Reviewer: I think we should drop the `all` here and in l.808 as it currently wouldn't be the case

Author: This proposal is aimed at future behaviour rather than current. The idea here is that the MHC would be defaulted to be in working order which (as far as I understand) is not the current behaviour.

Author: I've removed the 'all' in both places, but I'm still unclear on what difference there is in meaning.

Reviewer: Slight difference is that some fields are defaulted but not all (UnhealthyConditions). Of course pending discussion the issue about defaulting.

                apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
                kind: VSphereMachineTemplate
                name: windows-vsphere-template

>           # This will create a health check for each deployment created with the "windows-worker" MachineDeploymentClass

Reviewer: ```suggestion
          # This will create a MachineHealthCheck for each deployment created with the "windows-worker" MachineDeploymentClass
```
nit

    1. Creates the Machine Deployment object in the API server.
6. MachineHealthCheck object creation
   1. If `ClusterClass.spec.machineHealthCheck` is set:
      1. For each `Cluster.spec.topology.workers.machineDeployments` item in the list create a MachineHealthCheck with the label `topology.cluster.x-k8s.io/deployment-name: <machine-deployment-topology-name>` as the selector and set the `clusterName` field from the cluster. 
      2. For the Cluster.spec.topology.controlPlane create a MachineHealthCheck with the label `cluster.x-k8s.io/control-plane` as the selector and set the `clusterName` field from the cluster.
   2. For each `Cluster.spec.topology.workers.machineDeployments` with `ClusterClass.spec.workers.machineDeployments[i].machineDeploymentClass` defined create a MachineHealthCheck with the label `topology.cluster.x-k8s.io/deployment-name: <machine-deployment-topology-name>` as the selector and set the `clusterName` field from the cluster.
   3. If `Cluster.spec.topology.controlPlane.machineHealthCheckClass` is set create a MachineHealthCheck with the label `cluster.x-k8s.io/control-plane` as the selector and set the `clusterName` field from the cluster.

> **Note** `ClusterClass.spec.workers.machineDeployments[i].machineDeploymentClass` and `Cluster.spec.topology.controlPlane.machineHealthCheckClass` can not be set if `Cluster.spec.topology.workers.machineDeployments` is set.

Reviewer: Maybe we should follow the same structure as above with the InfraCluster/ControlPlane/MD creation or even embed it in that logic that it would only be an additional step during  InfraCluster/ControlPlane/MD reconciliation

> **Note** `ClusterClass.spec.workers.machineDeployments[i].machineDeploymentClass` and `Cluster.spec.topology.controlPlane.machineHealthCheckClass` can not be set if `Cluster.spec.topology.workers.machineDeployments` is set.

Author: I've add the If NOT line as the simpler of the two options - let me know if this works.

> **Note** `ClusterClass.spec.workers.machineDeployments[i].machineDeploymentClass` and `Cluster.spec.topology.controlPlane.machineHealthCheckClass` can not be set if `Cluster.spec.topology.workers.machineDeployments` is set.

Reviewer: Fine for me!
